lnrpc/routerrpc/router_backend.go

```diff
@@ -321,3 +340,166 @@ func (r *RouterBackend) UnmarshallRoute(rpcroute *lnrpc.Route) (
+			return nil, err
+		}
+
+		// Next, we'll ensure that this payreq hasn't already expired.
+		err = validatePayReqExpiry(payReq)
+		if err != nil {
+			return nil, err
+		}
+
+		// If the amount was not included in the invoice, then we let
+		// the payee specify the amount of satoshis they wish to send.
+		// We override the amount to pay with the amount provided from
+		// the payment request.
+		if payReq.MilliSat == nil {
+			if rpcPayReq.Amt == 0 {
+				return nil, errors.New("amount must be " +
+					"specified when paying a zero amount " +
+					"invoice")
+			}
+
+			payIntent.Amount = lnwire.NewMSatFromSatoshis(
+				btcutil.Amount(rpcPayReq.Amt),
+			)
+		} else {
+			if rpcPayReq.Amt != 0 {
+				return nil, errors.New("amount must not be " +
+					"specified when paying a non-zero " +
+					" amount invoice")
+			}
+
+			payIntent.Amount = *payReq.MilliSat
+		}
+
+		copy(payIntent.PaymentHash[:], payReq.PaymentHash[:])
+		destKey := payReq.Destination.SerializeCompressed()
+		copy(payIntent.Target[:], destKey)
+
+		payIntent.FinalCLTVDelta = int16(payReq.MinFinalCLTVExpiry())
+		payIntent.RouteHints = payReq.RouteHints
+	} else {
+		// Otherwise, If the payment request field was not specified
+		// (and a custom route wasn't specified), construct the payment
+		// from the other fields.
+
+		// Payment destination.
+		if len(rpcPayReq.Dest) != 33 {
+			return nil, errors.New("invalid key length")
+
+		}
+		pubBytes := rpcPayReq.Dest
+		copy(payIntent.Target[:], pubBytes)
+
+		// Final payment CLTV delta.
+		if rpcPayReq.FinalCltvDelta != 0 {
+			payIntent.FinalCLTVDelta =
+				int16(rpcPayReq.FinalCltvDelta)
+		} else {
+			payIntent.FinalCLTVDelta = zpay32.DefaultFinalCLTVDelta
+		}
+
+		// Amount.
+		if rpcPayReq.Amt == 0 {
+			return nil, errors.New("amount must be specified")
+		}
+
+		payIntent.Amount = lnwire.NewMSatFromSatoshis(
+			btcutil.Amount(rpcPayReq.Amt),
+		)
+
+		// If we're in debug HTLC mode, then all outgoing HTLCs will pay
+		// to the same debug rHash. Otherwise, we pay to the rHash
+		// specified within the RPC request.
+		if r.DebugHTLC &&
+			bytes.Equal(payIntent.PaymentHash[:], zeroHash[:]) {
+
+			copy(payIntent.PaymentHash[:], invoices.DebugHash[:])
+		} else {
+			copy(payIntent.PaymentHash[:], rpcPayReq.PaymentHash)
+		}
+	}
+
+	// Currently, within the bootstrap phase of the network, we limit the
+	// largest payment size allotted to (2^32) - 1 mSAT or 4.29 million
+	// satoshis.
+	if payIntent.Amount > r.MaxPaymentMSat {
+		// In this case, we'll send an error to the caller, but
+		// continue our loop for the next payment.
+		return payIntent, fmt.Errorf("payment of %v is too large, "+
+			"max payment allowed is %v", payIntent.Amount,
+			r.MaxPaymentMSat)
+
+	}
+
+	return payIntent, nil
+}
+
+// validatePayReqExpiry checks if the passed payment request has expired. In
+// the case it has expired, an error will be returned.
+func validatePayReqExpiry(payReq *zpay32.Invoice) error {
```

Comment:
export s.t. can be used by the rpcserver also?

Reply:
exported and reused